packages/grafana-ui/src/themes/ThemeContext.tsx

```diff
@@ -98,23 +100,35 @@ export function useStyles<T>(getStyles: (theme: GrafanaTheme) => T) {
 }
 
 /**
- * Hook for using memoized styles with access to the theme.
+ * Hook for using memoized styles with access to the theme. Pass additional
+ * arguments to the getStyles function as additional arguments to this hook.
  *
- * NOTE: For memoization to work, you need to ensure that the function
- * you pass in doesn't change, or only if it needs to. (i.e. declare
- * your style creator outside of a function component or use `useCallback()`.)
+ * Prefer using primitive values (boolean, number, string, etc) for
+ * additional arguments for better performance
+ *
+ * const getStyles = (theme, isDisabled, isOdd) => {css(...)}
+ * [...]
+ * const styles = useStyles2(getStyles, true, Boolean(index % 2))
+ *
+ * NOTE: For memoization to work, ensure that all arguments don't change
+ * across renders (or only change if they need to)
  * */
 /** @public */
-export function useStyles2<T>(getStyles: (theme: GrafanaTheme2) => T) {
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function useStyles2<T extends any[], CSSReturnValue>(
+  getStyles: (theme: GrafanaTheme2, ...args: T) => CSSReturnValue,
+  ...additionalArguments: T
+): CSSReturnValue {
   const theme = useTheme2();
 
   let memoizedStyleCreator: typeof getStyles = memoizedStyleCreators.get(getStyles);
+
   if (!memoizedStyleCreator) {
-    memoizedStyleCreator = stylesFactory(getStyles);
+    memoizedStyleCreator = memoize(getStyles, { maxSize: 10 }); // each getStyles function will memoize 10 different sets of props
```

Comment:
We can't use stylesFactory/memoize-one because we want to cache more than one call per getStyles function (the `isOdd` in a loop use case). 10 is an arbitary number (i put little thought into it - we could make it larger or smaller), and it will probably be unsuitable for primitive grafana-ui components like stack/box.